contracts/savings/SavingsManager.sol

```diff
@@ -140,32 +180,98 @@ contract SavingsManager is ISavingsManager, PausableModule {
 
+        uint256 leftover = _allUnclaimedRewards(_mAsset);
+        _initialiseStream(_mAsset, _liquidated.add(leftover), DURATION);
+
+        emit LiquidatorDeposited(_mAsset, _liquidated);
+    }
+
+    /**
+     * @dev Collects the platform interest from a given mAsset and then adds capital to the
+     * stream. If there is > 24h left in current stream, just top it up, otherwise reset.
+     * @param _mAsset The mAsset to fetch interest
+     */
+    function collectAndStreamInterest(address _mAsset)
+        external
+        whenNotPaused
+        whenStreamsNotFrozen
+    {
+        ISavingsContract savingsContract = savingsContracts[_mAsset];
+        require(address(savingsContract) != address(0), "Must have a valid savings contract");
+
         uint256 currentTime = now;
+        uint256 previousBatch = lastBatchCollected[_mAsset];
+        uint256 timeSincePreviousBatch = currentTime.sub(previousBatch);
+        require(timeSincePreviousBatch > 12 hours, "Cannot deposit twice in 12 hours");
+        lastBatchCollected[_mAsset] = currentTime;
 
-        // Get remaining rewards
-        uint256 end = rewardEnd[_mAsset];
+        // Batch collect
+        IMasset mAsset = IMasset(_mAsset);
+        (uint256 interestCollected, uint256 totalSupply) = mAsset.collectPlatformInterest();
+
+        if(interestCollected > 0){
+            // Validate APY
+            uint256 apy = _validateCollection(totalSupply, interestCollected, timeSincePreviousBatch);
+
+            // Get remaining rewards
+            uint256 leftover = _allUnclaimedRewards(_mAsset);
+
+            // Bundle this yield in with liquidator yield
+            // If (remainingTime < 24h)
+            //       take all remaining, add this, stream over next 24h
+            // else stream over remainingTime
+            uint256 end = streamEnd[_mAsset];
+            uint256 remaining = end > currentTime ? end.sub(currentTime) : 0;
+            uint256 newDuration = remaining < ONE_DAY ? ONE_DAY : remaining;
+            _initialiseStream(_mAsset, interestCollected.add(leftover), newDuration);
+
+            emit InterestCollected(_mAsset, interestCollected, totalSupply, apy);
+        }
+
+        emit InterestCollected(_mAsset, interestCollected, totalSupply, 0);
+    }
+
+    /**
+     * @dev Calculates how many rewards from the stream are still to be distributed, from the
+     * last collection time to the end of the stream.
+     * @param _mAsset The mAsset in question
+     * @return leftover The total amount of mAsset that is yet to be collected from a stream
+     */
+    function _allUnclaimedRewards(address _mAsset) internal view returns (uint256 leftover) {
+        uint256 currentTime = now;
+
+        uint256 end = streamEnd[_mAsset];
         uint256 lastUpdate = lastCollection[_mAsset];
         uint256 unclaimedSeconds = 0;
         if(currentTime <= end || lastUpdate < end){
```

Comment:
What's the reason for the disjunct `currentTime <= end`?

Reply:
Good catch.. this may have been an oversight from when i took inspiration from line 373. Current time can't be < lastUpdate, therefore this is redundant.